File: app/src/main/java/com/xpf/p2p/activity/AccountSafeActivity.java
package com.xpf.p2p.activity;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.text.TextUtils;
import android.view.View;
import android.widget.Button;
import android.widget.CompoundButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;
import android.widget.ToggleButton;

import com.xpf.p2p.R;
import com.xpf.p2p.base.BaseActivity;

/**
 * Created by x-sir on 2016/8/3 :)
 * Function:账户安全页面
 * {@link # https://github.com/xinpengfei520/P2P}
 */
public class AccountSafeActivity extends BaseActivity {

    ImageView ivBack;
    TextView tvTitle;
    ImageView ivSetting;
    ToggleButton togglebutton;
    Button btnResetGesture;
    LinearLayout activityAccountSafe;

    private SharedPreferences sp;

    @Override
    protected void initData() {
        ivBack = (ImageView) findViewById(R.id.iv_back);
        tvTitle = (TextView) findViewById(R.id.tv_title);
        ivSetting = (ImageView) findViewById(R.id.iv_setting);
        togglebutton = (ToggleButton) findViewById(R.id.togglebutton);
        btnResetGesture = (Button) findViewById(R.id.btn_resetGesture);
        activityAccountSafe = (LinearLayout) findViewById(R.id.activity_account_safe);
        ivBack.setVisibility(View.VISIBLE);
        ivSetting.setVisibility(View.GONE);
        tvTitle.setText("账户安全设置");

        ivBack.setOnClickListener(v -> removeCurrentActivity());
        btnResetGesture.setOnClickListener(v -> {
            // 给页面中的“button”设置重置密码的点击事件
            goToActivity(GestureEditActivity.class, null);
        });

        sp = this.getSharedPreferences("secret_protect", Context.MODE_PRIVATE);
        // 读取当前的toggleButton的状态并显示
        boolean isOpen = sp.getBoolean("isOpen", false);
        togglebutton.setChecked(isOpen);
        btnResetGesture.setEnabled(isOpen);//设置button的可操作性

        togglebutton.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                if (isChecked) {

                    // 判断本地是否设置过手势密码
                    String inputCode = sp.getString("inputCode", "");
                    if (TextUtils.isEmpty(inputCode)) {

                        Toast.makeText(AccountSafeActivity.this, "开启手势解锁", Toast.LENGTH_SHORT).show();
                        new AlertDialog.Builder(AccountSafeActivity.this)
                                .setTitle("是否现在设置手势密码")
                                .setPositiveButton("确定", new DialogInterface.OnClickListener() {
                                    @Override
                                    public void onClick(DialogInterface dialog, int which) {
                                        String inputCode1 = sp.getString("inputCode", "");
                                        if (TextUtils.isEmpty(inputCode1)) { // 用户之前没有设置过
                                            Toast.makeText(AccountSafeActivity.this, "现在设置手势密码", Toast.LENGTH_SHORT).show();
                                            // toggleButton.setChecked(true);
                                            btnResetGesture.setEnabled(true);
                                            sp.edit().putBoolean("isOpen", true).commit();
                                            goToActivity(GestureEditActivity.class, null);
                                        }
                                    }
                                })
                                .setNegativeButton("取消", new DialogInterface.OnClickListener() {
                                    @Override
                                    public void onClick(DialogInterface dialog, int which) {
                                        Toast.makeText(AccountSafeActivity.this, "取消设置手势密码", Toast.LENGTH_SHORT).show();
                                        togglebutton.setChecked(false);
                                        btnResetGesture.setEnabled(false);
                                        sp.edit().putBoolean("isOpen", false).commit();
                                    }
                                }).show();
                    } else { // 之前设置过
                        Toast.makeText(AccountSafeActivity.this, "开启了手势密码", Toast.LENGTH_SHORT).show();
                        // toggleButton.setChecked(true);
                        btnResetGesture.setEnabled(true);
                        sp.edit().putBoolean("isOpen", true).commit();
                    }

                } else {
                    Toast.makeText(AccountSafeActivity.this, "关闭手势解锁", Toast.LENGTH_SHORT).show();
                    // toggleButton.setChecked(false);
                    sp.edit().putBoolean("isOpen", false).commit();
                    btnResetGesture.setEnabled(false);
                }
            }
        });

    }

    @Override
    protected int getLayoutId() {
        return R.layout.activity_account_safe;
    }
}

File: app/src/main/java/com/xpf/p2p/activity/AccountSafeActivity.kt
package com.xpf.p2p.activity

import android.app.AlertDialog
import android.content.Context
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.widget.Toast
import com.xpf.p2p.base.BaseActivity
import com.xpf.p2p.databinding.ActivityAccountSafeBinding

/**
 * Created by x-sir on 2016/8/3 :)
 * Function:账户安全页面
 */
class AccountSafeActivity : BaseActivity<ActivityAccountSafeBinding>() {

    override fun createViewBinding(inflater: LayoutInflater) = ActivityAccountSafeBinding.inflate(inflater)

    override fun initData() {
        binding.titleBar.ivBack.visibility = View.VISIBLE
        binding.titleBar.ivSetting.visibility = View.GONE
        binding.titleBar.tvTitle.text = "账户安全设置"

        binding.titleBar.ivBack.setOnClickListener { removeCurrentActivity() }
        binding.btnResetGesture.setOnClickListener {
            goToActivity(GestureEditActivity::class.java, null)
        }

        val sp = getSharedPreferences("secret_protect", Context.MODE_PRIVATE)
        val isOpen = sp.getBoolean("isOpen", false)
        binding.togglebutton.isChecked = isOpen
        binding.btnResetGesture.isEnabled = isOpen

        binding.togglebutton.setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) {
                val inputCode = sp.getString("inputCode", "")
                if (TextUtils.isEmpty(inputCode)) {
                    Toast.makeText(this, "开启手势解锁", Toast.LENGTH_SHORT).show()
                    AlertDialog.Builder(this)
                        .setTitle("是否现在设置手势密码")
                        .setPositiveButton("确定") { _, _ ->
                            val inputCode1 = sp.getString("inputCode", "")
                            if (TextUtils.isEmpty(inputCode1)) {
                                Toast.makeText(this, "现在设置手势密码", Toast.LENGTH_SHORT).show()
                                binding.btnResetGesture.isEnabled = true
                                sp.edit().putBoolean("isOpen", true).apply()
                                goToActivity(GestureEditActivity::class.java, null)
                            }
                        }
                        .setNegativeButton("取消") { _, _ ->
                            Toast.makeText(this, "取消设置手势密码", Toast.LENGTH_SHORT).show()
                            binding.togglebutton.isChecked = false
                            binding.btnResetGesture.isEnabled = false
                            sp.edit().putBoolean("isOpen", false).apply()
                        }.show()
                } else {
                    Toast.makeText(this, "开启了手势密码", Toast.LENGTH_SHORT).show()
                    binding.btnResetGesture.isEnabled = true
                    sp.edit().putBoolean("isOpen", true).apply()
                }
            } else {
                Toast.makeText(this, "关闭手势解锁", Toast.LENGTH_SHORT).show()
                sp.edit().putBoolean("isOpen", false).apply()
                binding.btnResetGesture.isEnabled = false
            }
        }
    }
}
